Refactor tests

nwu63 opened the pull request:
## Purpose
This PR refactors the existing tests by introducing a test class that handles various shared tasks
- run an optimization
- check that solutions are close
- check that the final inform is correct
- check the history file contains expected contents
- check hot start

I also added parameterized tests etc. for efficiency.

Not all existing tests use this class however:
- the two SNOPT test files can probably use this if we want, but will require a bit more work
- for some reason NSGA2 will segfault if the `objfunc` is defined within the class. Absolutely no idea why..

I'll keep this as a draft PR for now. The main thing I want to check is that we're not losing any coverage through this, so if reviewers can also check against the old tests and see if I missed anything. Finally, the way the tests are structured can probably improved, so any suggestions are welcome.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [x] Maintenance update
- [ ] Other (please describe)

sseraj reviewed (commented):
This is a major improvement! I don't see any obvious reduction in coverage other than SNOPT versions prior to 7.7.7. Is that intentional?

marcomangano commented on test/test_large_snopt.py:
```diff
@@ -59,83 +67,68 @@ def sens(self, xdict, funcs):
+        sol = self.optimize()
+        self.assert_solution_allclose(sol, 1e-5, partial_x=True)
+
+    def test_sparse_IPOPT(self):
```
yeah I agree, can't we use @parameterized(["SNOPT", "IPOPT"]) for `test_sparse` @nwu63 ?

nwu63 replied:
Good point, will rename the file and parameterize those. Also Marco because you didn't wrap that in quotes you just notified some random person on GitHub :stuck_out_tongue: 

I added IPOPT because I don't think we have been testing its sparse Jacobian capabilities, but maybe the problem size is a bit unreasonable. Should we reduce the size?

sseraj replied:
I think the size could be reduced. I just checked the Azure logs and this one test is about 50% of the 'Run Tests' time.

nwu63 replied:
For some reason, setting it to `5000` resulted in some segfaults within IPOPT, while 10000 is okay. Could you double check and see if you could reproduce that on your local machine?

sseraj replied:
Yes, I was able to reproduce the segfault. That is very strange.

nwu63 commented on test/test_hs015.py:
```diff
@@ -53,190 +78,32 @@ def sens(self, xdict, funcs):
-        assert_allclose(sol_xvars, dv["xvars"], atol=tol, rtol=tol)
-        # we check either optimum via try/except
-        try:
-            if optName == "SNOPT" and opt.version != "7.7.7":
```
Good point, will add

nwu63 commented on test/test_hs015.py:
```diff
@@ -249,34 +116,20 @@ def test_snopt(self):
-        self.optimize_with_hotstart("PSQP", 5e-12, optOptions=optOptions)
+        # dv = sol.getDVs()
+        # sol_xvars = [sol.variables["xvars"][i].value for i in range(2)]
+        # assert_allclose(sol_xvars, dv["xvars"], atol=tol, rtol=tol)
```
Ah, right these were testing `sol.variables` etc. I will try to generalize these and make sure the Solution object contained the right values in different attributes.

nwu63 replied:
I added some tests, but not all since it's a bit harder to generalize. I've opened #256 to address this, among other things.

marcomangano commented on test/testing_utils.py:
```diff
@@ -25,12 +28,35 @@ def assert_dict_allclose(actual, desired, atol=DEFAULT_TOL, rtol=DEFAULT_TOL, pa
-    for key in actual.keys():
-        if key in desired.keys() or not partial:
-            assert_allclose(actual[key], desired[key], atol=atol, rtol=rtol, err_msg=f"Failed for key {key}")
+        assert set(actual.keys()) == set(desired.keys())
```
why did you drop `assertEqual` for this check?

nwu63 replied:
`assertEqual` cannot be used because this function is not inside a `unittest.TestCase` class.

marcomangano replied:
Makes sense, did not see that

marcomangano reviewed (commented):
Altright, I think I listed all the remaining comments I had. Note that some of them are not necessarily related to your edits, but additional comments/info we could add during this refactoring.

This is a substantial PR and some commits are pretty big and hard to follow, so for future PRs I would suggest to either break them further or, more in general, add a comment to the commit message to clarify the rationale behind it?

sseraj commented on test/testing_utils.py:
```diff
@@ -0,0 +1,411 @@
+        # now we assert against the closest solution
+        # objective
+        # sol.fStar was broken for earlier versions of SNOPT
+        if self.optName == "SNOPT" and Version(self.optVersion) >= Version("7.7.7"):
```
What does `Version` do here? It introduces an additional dependency on `packaging` and I'm not sure it's necessary.

nwu63 replied:
It allows for "correct" comparison of version strings, which is a non-trivial task (even though it's relatively simple here). `packaging` is a dependency of `setuptools` I believe, but I've switched to using `pkg_resources` which is something built into `setuptools` and therefore should available for everyone.

sseraj replied:
Thanks

sseraj reviewed (commented):
This looks good to me other than one thing

nwu63 commented on test/testing_utils.py:
```diff
@@ -0,0 +1,411 @@
+
+
+def get_dict_distance(d, d2):
+    """This converts a flat 1D dict to array using sorted keys, then computes the L2 distance"""
```
Not sure what you mean here. The code is not 100% robust because the numpy array casts aren't 100% fool proof, but the approach itself should be valid no?

marcomangano replied:
Yeah I was just thinking about cases where the L2 norm would not be the the best approach but I think for testing specifically this is not an issue. Good to go!

nwu63 commented on test/test_tp109.py:
```diff
@@ -164,59 +167,28 @@ def test_snopt(self):
 
         # Check Solution
-        assert_allclose(sol2.objectives["obj"].value, 0.536206927538e04, atol=1e-2, rtol=1e-2)
+        self.assert_solution(sol2, 1e-2)
```
Hmm good point, I think the idea is that this is loose because SNOPT may not be available, so then we just assert the ALPSO answer. I will fix this.

nwu63 replied:
Actually ALPSO seems...kinda broken. I'll open an issue but I prefer not fixing this in this PR.

nwu63 commented on test/test_tp109.py:
```diff
@@ -164,59 +167,28 @@ def test_snopt(self):
+        sol1 = self.optimize(sens="CS")
 
-        sol2 = opt2(sol1)
+        # Now re-optimize with SNOPT
```
Yes, we run ALPSO to get near the optimum, then use SNOPT to "refine" by cold starting from the best ALPSO point. This PR does not change any of the tests, this is something that has been there for ages, and it's good because it tests that the `Solution` object can be used to directly cold start optimizations. Although we should check that the DVs get transferred I guess..

nwu63 commented on test/test_hs071.py:
```diff
@@ -37,81 +66,30 @@ def sens(self, xdict, funcs):
+        self.optName = "SLSQP"
+        histFileName = "hs071_SLSQP_setDV.hst"
         newDV = {"xvars": np.array([1, 4, 4, 1])}
-        optOptions = {"IFILE": "{}.out".format(test_name)}
```
The output file names are uniformly handled in the `OptTest` class, which are generated based on the name of the individual tests. See [here](https://github.com/mdolab/pyoptsparse/blob/update-tests/test/testing_utils.py#L114-L125) for the mapping of option name and suffixes, and [here](https://github.com/mdolab/pyoptsparse/blob/update-tests/test/testing_utils.py#L203-L221) for how the options are set.

marcomangano replied:
Oh gotcha, I saw that but removed the info right away

nwu63 commented on test/testing_utils.py:
```diff
@@ -0,0 +1,411 @@
+        """
+        optionName = OUTPUT_FILENAMES[self.optName]
+        for optionName, suffix in OUTPUT_FILENAMES[self.optName].items():
+            optOptions[optionName] = f"{self.name}_{self._testMethodName}{suffix}"
```
Yeah I know what you mean, these are coming from `unittest`. Will do.

nwu63 replied:
I ended up changing these to `self.id()` since that's a public API so hopefully more stable. It's also more explicit even though it will generate some pretty long file names.

nwu63 commented on test/test_hs071.py:
```diff
@@ -75,12 +74,11 @@ def test_slsqp_setDV(self):
         """
-        test_name = "hs071_SLSQP_setDV"
-        histFileName = "{}.hst".format(test_name)
+        self.optName = "SLSQP"
```
I did it because I didn't want to pass the same variable around everywhere (it's needed in a lot of places). I think it's okay but would like to hear some feedback on this. @sseraj any thoughts?

Yeah `self.opt` created some obscure pickling error, took me a while to figure it out (still not really sure why that was the case). In any case storing the optimizer version was easy enough.

After looking at the code, I will add some more documentation and also set placeholder attributes in the class, otherwise it's unclear what attributes are needed to use `OptTest`.

sseraj replied:
I can see it being clearer to pass as an argument, but I think it's fine as is because you added the docstring and this is just used for testing.

nwu63 commented:
> This is a substantial PR and some commits are pretty big and hard to follow, so for future PRs I would suggest to either break them further or, more in general, add a comment to the commit message to clarify the rationale behind it?

Yeah sorry about that, I tried to break it up but as you can see, it's hard to do when I'm changing many things at once. Because I will be testing things as I develop, it's hard to keep a nice linear git history as well. I will try to make my commit messages better though, I don't think I was very good at that.

Anyways, I think I addressed most of the concerns.
